dbt_semantic_interfaces/dbt_semantic_interfaces/protocols/dimension.py

```diff
@@ -0,0 +1,48 @@
+from __future__ import annotations
+
+from enum import Enum
+from typing import Optional, Protocol
+
+from dbt_semantic_interfaces.objects.metadata import Metadata
+from dbt_semantic_interfaces.objects.time_granularity import TimeGranularity
+
+
+class DimensionType(Enum):
+    """Determines types of values expected of dimensions."""
+
+    CATEGORICAL = "categorical"
+    TIME = "time"
+
+
+class DimensionValidityParams(Protocol):
+    """Parameters identifying a given dimension as an entity for validity state
+
+    This construct is used for supporting SCD Type II tables, such as might be
+    created via dbt's snapshot feature, or generated via periodic loads from external
+    dimension data sources. In either of those cases, there is typically a time dimension
+    associated with the SCD data source that indicates the start and end times of a
+    validity window, where the dimension value is valid for any time within that range.
+    """
+
+    is_start: bool
+    is_end: bool
+
+
+class DimensionTypeParams(Protocol):
```

Comment:
Yet another option is to keep it as type params but then change the internals of this to be bound to type, so we'd have:

```
DimensionTypeParams(Protocol):
    categorical_params: Optional[CategoricalDimensionParams]
    time_params: Optional[TimeDimensionParams]
```

Then each sub-struct would have the parameters specific to the type. In this case, we actually wouldn't even include categorical_params, because there are none. This seems a bit silly to me, but it is more expressive and allows for extensions to other dimension types. 

The pattern here would be more useful in the case of things like metrics, where derived metrics and measure_proxy metrics could have very different parameter sets, and when we add something like conversion_metrics we'll need a place to put those parameters as well.

Reply:
What if instead we do...
```python
DimensionTypeParams = Union[CategoricalDimensionParams, TimeDimensionParams]

class Dimension(Protocol):
  ...
  type_params: Optional[DimensionTypeParams]
  partitioning_params: Optional[DimensionPartitioningParams]
  ...
```